File: python/mrd_stream_recon.py
import argparse
from typing import BinaryIO, Iterable, Union
import sys
import mrd
import numpy as np
from transform import k2i

def acquisition_reader(input: Iterable[mrd.StreamItem]) -> Iterable[mrd.Acquisition]:
    for item in input:
        _, item_data = item
        if not isinstance(item_data, mrd.Acquisition):
            continue
        yield item_data

def stream_item_sink(input: Iterable[Union[mrd.Acquisition, mrd.Image[np.float32]]]) -> Iterable[mrd.StreamItem]:
    for item in input:
        if isinstance(item, mrd.Acquisition):
            yield ('Acquisition', item)
        elif isinstance(item, mrd.Image) and item.data.dtype == np.float32:
            yield ('ImageFloat', item)
        else:
            raise ValueError("Unknown item type")

def remove_oversampling(head: mrd.Header,input: Iterable[mrd.Acquisition]) -> Iterable[mrd.Acquisition]:
    enc = head.encoding[0]

    if enc.encoded_space and enc.encoded_space.matrix_size and enc.recon_space and enc.recon_space.matrix_size:
        eNx = enc.encoded_space.matrix_size.x
        rNx = enc.recon_space.matrix_size.x
    else:
        raise Exception('Encoding information missing from header')

    for acq in input:
        if eNx != rNx and acq.samples() == eNx:
            xline = k2i(acq.data, [1])
            x0 = int((eNx - rNx) / 2)
            x1 = int((eNx - rNx) / 2 + rNx)
            xline = xline[:, x0:x1]
            acq.center_sample = int(rNx/2)
            acq.data = xline.astype('complex64')  # Keep it in image space

        yield acq

def accumulate_fft(head: mrd.Header, input: Iterable[mrd.Acquisition]) -> Iterable[mrd.Image[np.float32]]:
    enc = head.encoding[0]

    # Matrix size
    if enc.encoded_space and enc.recon_space and enc.encoded_space.matrix_size and enc.recon_space.matrix_size:
        eNx = enc.encoded_space.matrix_size.x
        eNy = enc.encoded_space.matrix_size.y
        eNz = enc.encoded_space.matrix_size.z
        rNx = enc.recon_space.matrix_size.x
        rNy = enc.recon_space.matrix_size.y
        rNz = enc.recon_space.matrix_size.z
    else:
        raise Exception('Required encoding information not found in header')

    # Field of view
    if enc.recon_space and enc.recon_space.field_of_view_mm:
        rFOVx = enc.recon_space.field_of_view_mm.x
        rFOVy = enc.recon_space.field_of_view_mm.y
        rFOVz = enc.recon_space.field_of_view_mm.z if enc.recon_space.field_of_view_mm.z else 1
    else:
        raise Exception('Required field of view information not found in header')

    # Number of Slices, Reps, Contrasts, etc.
    ncoils = 1
    if head.acquisition_system_information and head.acquisition_system_information.receiver_channels:
        ncoils = head.acquisition_system_information.receiver_channels

    if enc.encoding_limits and enc.encoding_limits.slice != None:
        nslices = enc.encoding_limits.slice.maximum + 1
    else:
        nslices = 1

    ncontrasts = 1
    if enc.encoding_limits and enc.encoding_limits.contrast != None:
        ncontrasts = enc.encoding_limits.contrast.maximum + 1

    if enc.encoding_limits and enc.encoding_limits.kspace_encoding_step_1 != None:
        ky_offset = int((eNy+1)/2) - enc.encoding_limits.kspace_encoding_step_1.center
    else:
        ky_offset = 0

    current_rep = -1
    reference_acquisition = None
    buffer = None

    def produce_image(buffer: np.ndarray, ref_acq: mrd.Acquisition) -> Iterable[mrd.Image[np.float32]]:
        if buffer.shape[-3] > 1:
            img = k2i(buffer, dim=[-2, -3])  # Assuming FFT in x-direction has happened upstream
        else:
            img = k2i(buffer, dim=[-2])  # Assuming FFT in x-direction has happened upstream

        for contrast in range(img.shape[0]):
            for islice in range(img.shape[1]):
                img_combined = np.squeeze(np.sqrt(np.abs(np.sum(img[contrast, islice] * np.conj(img[contrast, islice]), axis=0)).astype('float32')))

                xoffset = int((img_combined.shape[0] + 1)/2) - int((rNx+1)/2)
                yoffset = int((img_combined.shape[1] + 1)/2) - int((rNy+1)/2)
                if len(img_combined.shape) == 3:
                    zoffset = int((img_combined.shape[2] + 1)/2) - int((rNz+1)/2)
                    img_combined = img_combined[xoffset:(xoffset+rNx), yoffset:(yoffset+rNy), zoffset:(zoffset+rNz)]
                elif len(img_combined.shape) == 2:
                    img_combined = img_combined[xoffset:(xoffset+rNx), yoffset:(yoffset+rNy)]
                else:
                    raise Exception('Array img_bytes should have 2 or 3 dimensions')
                img_combined = np.reshape(img_combined, (1,1,img_combined.shape[-2], img_combined.shape[-1]))
                mrd_image = mrd.Image[np.float32](image_type=mrd.ImageType.MAGNITUDE, data=img_combined)
                mrd_image.field_of_view[0] = rFOVx
                mrd_image.field_of_view[1] = rFOVy
                mrd_image.field_of_view[2] = rFOVz/rNz
                mrd_image.position = ref_acq.position
                mrd_image.col_dir = ref_acq.read_dir
                mrd_image.line_dir = ref_acq.phase_dir
                mrd_image.slice_dir = ref_acq.slice_dir
                mrd_image.patient_table_position = ref_acq.patient_table_position
                mrd_image.acquisition_time_stamp = ref_acq.acquisition_time_stamp
                mrd_image.physiology_time_stamp = np.array(ref_acq.physiology_time_stamp).astype('uint32')
                mrd_image.slice = ref_acq.idx.slice
                mrd_image.contrast = contrast
                mrd_image.repetition = ref_acq.idx.repetition
                mrd_image.phase = ref_acq.idx.phase
                mrd_image.average = ref_acq.idx.average
                mrd_image.set = ref_acq.idx.set
                yield mrd_image

    for acq in input:
        if acq.idx.repetition != current_rep:
            # If we have a current buffer pass it on
            if buffer is not None and reference_acquisition is not None:
                yield from produce_image(buffer, reference_acquisition)

            # Reset buffer
            if acq.data.shape[-1] == eNx:
                readout_length = eNx
            else:
                readout_length = rNx  # Readout oversampling has probably been removed upstream

            buffer = np.zeros((ncontrasts, nslices, ncoils, eNz, eNy, readout_length), dtype=np.complex64)
            current_rep = acq.idx.repetition
            reference_acquisition = acq

        # Stuff into the buffer
        if buffer is not None:
            contrast = acq.idx.contrast if acq.idx.contrast is not None else 0
            slice = acq.idx.slice if acq.idx.slice is not None else 0
            k1 = acq.idx.kspace_encode_step_1 if acq.idx.kspace_encode_step_1 is not None else 0
            k2 = acq.idx.kspace_encode_step_2 if acq.idx.kspace_encode_step_2 is not None else 0
            buffer[contrast, slice, :, k2, k1 + ky_offset, :] = acq.data

    if buffer is not None and reference_acquisition is not None:
        yield from produce_image(buffer, reference_acquisition)
        buffer = None
        reference_acquisition = None

def reconstruct_mrd_stream(input: BinaryIO, output: BinaryIO):
    with mrd.BinaryMrdReader(input) as reader:
        with mrd.BinaryMrdWriter(output) as writer:
            head = reader.read_header()
            if head is None:
                raise Exception("Could not read header")
            writer.write_header(head)
            writer.write_data(
                stream_item_sink(
                    accumulate_fft(head,
                        remove_oversampling(head,
                            acquisition_reader(reader.read_data())))))

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Reconstructs an MRD stream")
    parser.add_argument('-i', '--input', type=str, required=False, help="Input file, defaults to stdin")
    parser.add_argument('-o', '--output', type=str, required=False, help="Output file, defaults to stdout")
    args = parser.parse_args()

    input = open(args.input, "rb") if args.input is not None else sys.stdin.buffer
    output = open(args.output, "wb") if args.output is not None else sys.stdout.buffer

    reconstruct_mrd_stream(input, output)

File: cpp/mrd-tools/mrd_stream_recon.cc
#include "mrd/binary/protocols.h"
#include "mrd/protocols.h"
#include "mrd/types.h"

#include <xtensor-fftw/basic.hpp>
#include <xtensor-fftw/helper.hpp>
#include <xtensor/xstrided_view.hpp>
#include <xtensor/xview.hpp>

xt::xtensor<std::complex<float>, 4> fftshift(xt::xtensor<std::complex<float>, 4> x) {
  return xt::roll(xt::roll(x, x.shape(3) / 2, 3), x.shape(2) / 2, 2);
}

void remove_oversampling(mrd::Acquisition& acq, const mrd::EncodingType& enc) {
  auto new_shape = acq.data.shape();

  auto eNx = enc.encoded_space.matrix_size.x;
  auto rNx = enc.recon_space.matrix_size.x;
  new_shape[1] = rNx;
  auto x_pad = (eNx - rNx) / 2;
  xt::xtensor<std::complex<float>, 2> new_data = xt::zeros<std::complex<float>>(new_shape);
  for (size_t c = 0; c < acq.Coils(); c++) {
    auto ft_line = xt::xarray<std::complex<float>>(xt::view(acq.data, c, xt::all()));
    ft_line = xt::fftw::fftshift(xt::fftw::ifft(xt::fftw::ifftshift(ft_line)));
    ft_line *= std::sqrt(1.0f * ft_line.size());
    ft_line = xt::view(ft_line, xt::range(x_pad, rNx + x_pad));
    ft_line = xt::fftw::fftshift(xt::fftw::fft(xt::fftw::ifftshift(ft_line)));
    ft_line /= std::sqrt(1.0f * ft_line.size());
    xt::view(new_data, c, xt::all()) = ft_line;
  }
  acq.data = new_data;
}

void print_usage(std::string program_name) {
  std::cerr << "Usage: " << program_name << std::endl;
  std::cerr << "  -i|--input   <input MRD stream> (default: stdin)" << std::endl;
  std::cerr << "  -o|--output  <output MRD stream> (default: stdout)" << std::endl;
  std::cerr << "  -h|--help" << std::endl;
}

int main(int argc, char** argv) {
  std::string input_path;
  std::string output_path;

  std::vector<std::string> args(argv, argv + argc);
  auto current_arg = args.begin() + 1;
  while (current_arg != args.end()) {
    if (*current_arg == "--help" || *current_arg == "-h") {
      print_usage(args[0]);
      return 0;
    } else if (*current_arg == "--input" || *current_arg == "-i") {
      current_arg++;
      if (current_arg == args.end()) {
        std::cerr << "Missing input file" << std::endl;
        print_usage(args[0]);
        return 1;
      }
      input_path = *current_arg;
      current_arg++;
    } else if (*current_arg == "--output" || *current_arg == "-o") {
      current_arg++;
      if (current_arg == args.end()) {
        std::cerr << "Missing output file" << std::endl;
        print_usage(args[0]);
        return 1;
      }
      output_path = *current_arg;
      current_arg++;
    } else {
      std::cerr << "Unknown argument: " << *current_arg << std::endl;
      print_usage(args[0]);
      return 1;
    }
  }

  std::unique_ptr<std::ifstream> input_file;
  if (!input_path.empty()) {
    input_file = std::make_unique<std::ifstream>(input_path, std::ios::binary | std::ios::in);
    if (!input_file->good()) {
      throw std::runtime_error("Failed to open input file for reading.");
    }
  }

  std::unique_ptr<std::ofstream> output_file;
  if (!output_path.empty()) {
    output_file = std::make_unique<std::ofstream>(output_path, std::ios::binary | std::ios::out);
    if (!output_file->good()) {
      throw std::runtime_error("Failed to open output file for writing.");
    }
  }

  mrd::binary::MrdReader r(input_path.empty() ? std::cin : *input_file);
  mrd::binary::MrdWriter w(output_path.empty() ? std::cout : *output_file);

  std::optional<mrd::Header> ho;
  r.ReadHeader(ho);
  if (!ho) {
    std::cerr << "Failed to read header" << std::endl;
    return 1;
  }

  auto h = ho.value();
  // Just copy the header
  w.WriteHeader(h);

  auto enc = h.encoding[0];

  auto eNx = enc.encoded_space.matrix_size.x;
  auto eNy = enc.encoded_space.matrix_size.y;
  auto eNz = enc.encoded_space.matrix_size.z;
  auto rNx = enc.recon_space.matrix_size.x;
  auto rNy = enc.recon_space.matrix_size.y;
  auto rNz = enc.recon_space.matrix_size.z;
  auto rFOVx = enc.recon_space.field_of_view_mm.x;
  auto rFOVy = enc.recon_space.field_of_view_mm.y;
  auto rFOVz = enc.recon_space.field_of_view_mm.z;
  if (rFOVz == 0) {
    rFOVz = 1;
  }

  uint32_t ncoils = 1;
  if (h.acquisition_system_information && h.acquisition_system_information->receiver_channels) {
    ncoils = h.acquisition_system_information->receiver_channels.value();
  }

  uint32_t nslices = 1;
  if (enc.encoding_limits.slice.has_value()) {
    nslices = enc.encoding_limits.slice->maximum + 1;
  }

  uint32_t ncontrasts = 1;
  if (enc.encoding_limits.contrast.has_value()) {
    ncontrasts = enc.encoding_limits.contrast->maximum + 1;
  }

  uint32_t ky_offset = 0;
  if (enc.encoding_limits.kspace_encoding_step_1.has_value()) {
    ky_offset = ((eNy + 1) / 2) - enc.encoding_limits.kspace_encoding_step_1->center;
  }

  // When we have aliased types, we will use that.
  mrd::StreamItem v;
  xt::xtensor<std::complex<float>, 6> buffer;
  mrd::Acquisition ref_acq;
  uint32_t image_index = 0;
  while (r.ReadData(v)) {
    if (!std::holds_alternative<mrd::Acquisition>(v)) {
      continue;
    }
    auto acq = std::get<mrd::Acquisition>(v);

    // Currently ignoring noise scans
    if (acq.head.flags.HasFlags(mrd::AcquisitionFlags::kIsNoiseMeasurement)) {
      continue;
    }

    // Remove oversampling
    if (eNx != rNx && acq.Samples() == eNx) {
      remove_oversampling(acq, enc);
    }

    // If this is the first line, we need to allocate the buffer
    if (acq.head.flags.HasFlags(mrd::AcquisitionFlags::kFirstInEncodeStep1) || acq.head.flags.HasFlags(mrd::AcquisitionFlags::kFirstInSlice)) {
      uint32_t readout_length = acq.Samples();
      std::array<size_t, 6> shape = {ncontrasts, nslices, ncoils, eNz, eNy, readout_length};
      buffer = xt::zeros<std::complex<float>>(shape);
      ref_acq = acq;
    }

    // Copy the data into the buffer
    auto contrast = acq.head.idx.contrast.value_or(0);
    auto slice = acq.head.idx.slice.value_or(0);
    auto k1 = acq.head.idx.kspace_encode_step_1.value_or(0);
    auto k2 = acq.head.idx.kspace_encode_step_2.value_or(0);
    xt::view(buffer, contrast, slice, xt::all(), k2, k1 + ky_offset, xt::all()) = xt::xarray<std::complex<float>>(acq.data);

    // If this is the last line, we need to write the buffer
    if (acq.head.flags.HasFlags(mrd::AcquisitionFlags::kLastInEncodeStep1) || acq.head.flags.HasFlags(mrd::AcquisitionFlags::kLastInSlice)) {
      for (unsigned int c = 0; c < buffer.shape(0); c++) {
        for (unsigned int s = 0; s < buffer.shape(1); s++) {
          auto slice = xt::view(buffer, c, s, xt::all(), xt::all(), xt::all(), xt::all());
          slice = fftshift(slice);
          for (unsigned int coil = 0; coil < slice.shape(0); coil++) {
            auto tmp1 = xt::fftw::ifft2(xt::xarray<std::complex<float>>(xt::view(slice, coil, 0, xt::all(), xt::all())));
            tmp1 *= std::sqrt(1.0f * tmp1.size());
            xt::view(slice, coil, 0, xt::all(), xt::all()) = tmp1;
          }
          slice = fftshift(slice);

          std::array<size_t, 4> image_shape = {1, slice.shape(1), slice.shape(2), slice.shape(3)};
          auto pixel_data = xt::sqrt(xt::abs(xt::sum(slice * xt::conj(slice), 0)));

          auto xoffset = (slice.shape(3) + 1) / 2 - (rNx + 1) / 2;
          auto yoffset = (slice.shape(2) + 1) / 2 - (rNy + 1) / 2;
          auto zoffset = (slice.shape(1) + 1) / 2 - (rNz + 1) / 2;
          auto combined = xt::view(pixel_data,
                                   xt::range(zoffset, zoffset + rNz),
                                   xt::range(yoffset, yoffset + rNy),
                                   xt::range(xoffset, xoffset + rNx));

          mrd::Image<float> im;
          im.data = xt::zeros<float>(image_shape);
          xt::view(im.data, 0, xt::all(), xt::all(), xt::all()) = combined;

          im.head.measurement_uid = acq.head.measurement_uid;
          im.head.field_of_view[0] = rFOVx;
          im.head.field_of_view[1] = rFOVy;
          im.head.field_of_view[2] = rFOVz;
          im.head.position = ref_acq.head.position;
          im.head.col_dir = ref_acq.head.read_dir;
          im.head.line_dir = ref_acq.head.phase_dir;
          im.head.slice_dir = ref_acq.head.slice_dir;
          im.head.patient_table_position = ref_acq.head.patient_table_position;
          im.head.average = ref_acq.head.idx.average;
          im.head.slice = ref_acq.head.idx.slice;
          im.head.contrast = ref_acq.head.idx.contrast;
          im.head.phase = ref_acq.head.idx.phase;
          im.head.repetition = ref_acq.head.idx.repetition;
          im.head.set = ref_acq.head.idx.set;
          im.head.acquisition_time_stamp_ns = ref_acq.head.acquisition_time_stamp_ns;
          im.head.physiology_time_stamp_ns = ref_acq.head.physiology_time_stamp_ns;
          im.head.image_type = mrd::ImageType::kMagnitude;
          im.head.image_index = image_index++;
          im.head.image_series_index = 0;
          im.head.user_int = ref_acq.head.user_int;
          im.head.user_float = ref_acq.head.user_float;
          w.WriteData(im);
        }
      }
    }
  }

  w.EndData();

  return 0;
}
